lib/thrift/servers/binary_framed/acceptor.ex

```diff
@@ -0,0 +1,73 @@
+defmodule Thrift.Servers.BinaryFramed.Acceptor do
+  @moduledoc """
+  A GenServer that accepts connections on a server and processes the thrift messages.
+  """
+
+  alias Thrift.{
+    Protocol,
+    TApplicationException
+  }
+  use GenServer
+  require Logger
+
+  @spec start_link(pid, module, module) :: GenServer.on_start
+  def start_link(socket, server_module, handler_module) do
+    GenServer.start_link(__MODULE__, [socket, server_module, handler_module])
+  end
+
+  def init([socket, server_module, handler_module]) do
+    Process.send_after(self, :accept, 1)
+    {:ok, {socket, server_module, handler_module}}
+  end
+
+  def handle_info(:accept, {socket, server_module, handler_module}) do
+    {:ok, r_sock} = :gen_tcp.accept(socket)
+
+    do_thrift_call({r_sock, server_module, handler_module})
+  end
+
+  defp do_thrift_call({socket, server_module, handler_module} = args) do
+    err = with {:ok, message}   <- :gen_tcp.recv(socket, 0),
+    parsed_response             <- Protocol.Binary.deserialize(:message_begin, message),
+    {:ok, thrift_data}          <- handle_thrift_message(parsed_response, server_module, handler_module) do
+
+      :gen_tcp.send(socket, thrift_data)
```

Comment:
Is there a way to structure this so that the `gen_tcp` calls are within the same scope?  I find it easier to reason about code if paired side effects are kept close together - i.e.,
```
input = receive_input(...)
output = do_stuff_with(input)  # <- purely functional if possible
send_output(output)
```